WikiWatershed/model-my-watershed
# Export Project to HydroShare

## Overview

This PR sets up backend and frontend components that allow the user to export a project to HydroShare. When a logged in user opens the MultiShare view in a saved project, they see the HydroShare Export switch. When this switch is turned on we first check to see if the user has authorized HydroShare. If not, we ask them to login and go through the OAuth authorization dance. With the authorization, we open a modal asking for the user to specify a title (defaulting to the project's name) and abstract, and optionally some keywords. When submitted, these values are combined with a new CSV export of each analysis type. Nothing from the modeling stage is currently included.

The above payload is sent to a new endpoint which verifies that the project exists and is owned by the user, then creates a resource in HydroShare with the given title, abstract, keywords, and files. It also adds a GeoJSON and Shapefile for area of interest of the project.

When this completes, the frontend updates to show a link to the project. Turning the switch off at this point opens a confirmation modal informing the user of the irreversibility of their action, pending which the resource is deleted from HydroShare.

This is a first draft, and there will be a number of follow up cards from this.

Connects #2544 
Connects #2557 

### Demo

![2017-12-27 15 52 40](https://user-images.githubusercontent.com/1430060/34393172-9338cf44-eb1e-11e7-8804-564e00aff6a1.gif)

### Notes

* We switch from using `dev.hydroshare` to `beta.hydroshare` on the suggestion of HydroShare developers. `beta` shares credentials with `production`, so our secrets file will have to be updated correspondingly.

* In case an analysis fails or is still running when the export is done, it will not be included in the export.

* The actual export takes quite some time: 10-15s. I'm not sure which step is the biggest bottleneck, I suspect it is the Shapefile generation, but haven't tested for sure. We can consider breaking this off into a Celery task with some persistent UI spinner to show this status outside of the Share modal.

* The CSVs generated for the Analyze tabs are markedly different from the ones retrieved through the UI. This is because the UI table CSVs work off rendered tables, however this is happening in the background and thus we work with raw model data. We can make these match in a future enhancement.

* A number of outstanding issues remain, which will be covered in later cards. A full list of these can be seen in #2507. Briefly, they are:

  * Group shapefiles in a folder for easy access
  * POST to `/export/hydroshare?project=XX` for an existing project should update that resource
  * Add UI for facilitating "export on save" so that every user action leads to a corresponding export
  * Analyze CSVs do not match those generated from the UI
  * Keywords should use bubbles in the UI
  * MapShed projects should export GMS files for every Scenario
  * Language and UI polish ("Unsychronize", error messages for edge cases)

  I'll be creating cards for these shortly.

## Testing Instructions

* Check out this branch, reprovision `app`, run migrations, and bundle:

      $ MMW_HYDROSHARE_SECRET_KEY=*** vagrant reload app --provision
      $ ./scripts/manage.sh migrate
      $ ./scripts/bundle.sh --debug

* Go to https://beta.hydroshare.org/ and log in with credentials in the `HydroShare OAuth MMW Beta` LastPass site.
* Go to [:8000/](http://localhost:8000/), log in, and create and save a project
* Go to the share menu, and enable HydroShare Export
* Authorize MMW
* In the Export modal, verify that the title matches the project name. Fill in an abstract, and optionally some keywords. Click "Export"
* Wait for the export to finish. When it does, there should be a link to the resource in HydroShare. Click it, and ensure it has everything advertised.
* Try downloading some files and ensure they work as expected.
* Go back to MMW, and switch off HydroShare. Once that completes, ensure that the resource has been deleted from HydroShare.
